code/modules/mechanics/MechanicsTextToMusicCom.dm

```diff
@@ -0,0 +1,359 @@
+		if (src.anchored)
+			src.ready_piano()
+
+	proc/mechcompNotes(var/datum/mechanicsMessage/input)
+		if (input.signal)
+			src.set_notes(input.signal)
+
+	proc/mechcompTiming(var/datum/mechanicsMessage/input)
+		var/new_timing = text2num(input.signal)
+		if (new_timing)
+			src.set_timing(new_timing)
+
+	proc/mechcompInstrument(var/datum/mechanicsMessage/input)
+		if (input.signal)
+			src.set_instrument(input.signal)
+
+	proc/mechcompStop(var/datum/mechanicsMessage/input)
+		if (src.is_busy)
+			src.is_stop_requested = TRUE
+
+	proc/mechcompReset(var/datum/mechanicsMessage/input)
+		src.reset_piano(0)
+
+	// ------------------------------------------------
+
+	proc/mechcompConfigPlay(obj/item/W as obj, mob/user as mob)
+		if (src.anchored)
+			src.ready_piano()
+
+	proc/mechcompConfigNotes(obj/item/W as obj, mob/user as mob)
+		var/given_notes = tgui_input_text(user, "Input notes to play.", "Set Notes", src.note_input)
+		src.set_notes(given_notes)
+
+	proc/mechcompConfigTiming(obj/item/W as obj, mob/user as mob)
+		// `tgui_input_number` behaves oddly when dealing with floats
+		// so I'm using the text input instead, unless timing is counted is centiseconds
+		var/new_timing = tgui_input_text(user, "Input a new timing between [MIN_TIMING] and [MAX_TIMING] seconds.", "Set Timing", src.timing)
+		new_timing = text2num(new_timing)
+		if (new_timing)
+			src.set_timing(new_timing)
+
+	proc/mechcompConfigInstrument(obj/item/W as obj, mob/user as mob)
+		var/new_instrument = tgui_input_list(user, "Input new instrument.", "Set Instrument", src.allow_list, src.instrument_sound_path)
+		src.set_instrument(new_instrument)
+
+	proc/mechcompConfigStop(obj/item/W as obj, mob/user as mob)
+		if (src.is_busy)
+			src.is_stop_requested = TRUE
+
+	proc/mechcompConfigReset(obj/item/W as obj, mob/user as mob)
+		src.reset_piano(0)
+
+	proc/mechcompConfigViewErrors(obj/item/W as obj, mob/user as mob)
+		if (length(src.error_messages) > 0)
+			var/message = jointext(src.error_messages, "<br><hr>")
+			tgui_message(user, message, "T2M Error Messages")
+
+	disposing() //just to clear up ANY funkiness
+		reset_piano(1)
+		..()
+
+	attackby(obj/item/W, mob/user)
+		if(iswrenchingtool(W) && src.anchored && src.is_busy)
+			src.is_stop_requested = TRUE
+
+		return ..()
+
+	proc/log_error_message(var/error_message)
+		src.error_messages.Insert(1, error_message)
+
+		if (length(src.error_messages) > MAX_ERROR_MESSAGES)
+			src.error_messages.Cut(MAX_ERROR_MESSAGES + 1, 0)
+
+		animate_flash_color_fill(src,"#ff0000",2, 2)
+
+	proc/clean_input() //breaks our big input string into chunks
+		src.is_busy = 1
+		src.piano_notes = list()
+		var/list/split_input = splittext("[note_input]", "|")
+		if (split_input > MAX_NOTE_INPUT)
+			return FALSE
+		for (var/string in split_input)
+			if (string)
+				src.piano_notes += string
+		src.is_busy = 0
+		return TRUE
+
+	proc/build_notes(var/list/piano_notes) //breaks our chunks apart and puts them into lists on the object
```

Comment:
having this much duplicated code between this and player pianos is a bit of a maintainability nightmare, and should probably be refactored